File: platforms/ios/SloppadIOS/SloppadModels.swift
import Foundation

struct SloppadLoginRequest: Encodable {
    let password: String
}

struct SloppadWorkspaceListResponse: Decodable {
    let ok: Bool
    let activeWorkspaceID: String
    let workspaces: [SloppadWorkspace]

    private enum CodingKeys: String, CodingKey {
        case ok
        case activeWorkspaceID = "active_workspace_id"
        case workspaces
    }
}

struct SloppadWorkspace: Decodable, Identifiable, Hashable {
    let id: String
    let name: String
    let rootPath: String
    let chatSessionID: String
    let canvasSessionID: String

    private enum CodingKeys: String, CodingKey {
        case id
        case name
        case rootPath = "root_path"
        case chatSessionID = "chat_session_id"
        case canvasSessionID = "canvas_session_id"
    }
}

struct SloppadChatHistoryResponse: Decodable {
    let messages: [SloppadPersistedMessage]
}

struct SloppadPersistedMessage: Decodable, Identifiable {
    let id: Int64
    let role: String
    let contentMarkdown: String
    let contentPlain: String

    private enum CodingKeys: String, CodingKey {
        case id
        case role
        case contentMarkdown = "content_markdown"
        case contentPlain = "content_plain"
    }

    var content: String {
        if contentMarkdown.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty == false {
            return contentMarkdown
        }
        return contentPlain
    }
}

struct SloppadChatSendRequest: Encodable {
    let text: String
    let outputMode: String

    private enum CodingKeys: String, CodingKey {
        case text
        case outputMode = "output_mode"
    }
}

struct SloppadRenderedMessage: Identifiable, Equatable {
    let id: String
    let role: String
    let text: String
    let html: String
}

struct SloppadCanvasArtifact: Equatable {
    let kind: String
    let title: String
    let html: String
    let text: String
}

struct SloppadCanvasSnapshotResponse: Decodable {
    let event: SloppadCanvasEventPayload?
}

struct SloppadCanvasEventPayload: Decodable {
    let kind: String?
    let title: String?
    let html: String?
    let text: String?
    let markdownOrText: String?
    let path: String?

    private enum CodingKeys: String, CodingKey {
        case kind
        case title
        case html
        case text
        case markdownOrText = "markdown_or_text"
        case path
    }
}

struct SloppadChatEventPayload: Decodable {
    let type: String
    let turnID: String?
    let role: String?
    let message: String?
    let markdown: String?
    let html: String?
    let error: String?
    let text: String?
    let reason: String?
    let state: String?
    let workspacePath: String?
    let actionType: String?

    private enum CodingKeys: String, CodingKey {
        case type
        case turnID = "turn_id"
        case role
        case message
        case markdown
        case html
        case error
        case text
        case reason
        case state
        case workspacePath = "workspace_path"
        case action
    }

    private struct ActionPayload: Decodable {
        let type: String?
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        type = try container.decode(String.self, forKey: .type)
        turnID = try container.decodeIfPresent(String.self, forKey: .turnID)
        role = try container.decodeIfPresent(String.self, forKey: .role)
        message = try container.decodeIfPresent(String.self, forKey: .message)
        markdown = try container.decodeIfPresent(String.self, forKey: .markdown)
        html = try container.decodeIfPresent(String.self, forKey: .html)
        error = try container.decodeIfPresent(String.self, forKey: .error)
        text = try container.decodeIfPresent(String.self, forKey: .text)
        reason = try container.decodeIfPresent(String.self, forKey: .reason)
        state = try container.decodeIfPresent(String.self, forKey: .state)
        workspacePath = try container.decodeIfPresent(String.self, forKey: .workspacePath)
        actionType = try container.decodeIfPresent(ActionPayload.self, forKey: .action)?.type
    }
}

struct SloppadCompanionConfig: Decodable {
    let companionEnabled: Bool
    let idleSurface: String

    private enum CodingKeys: String, CodingKey {
        case companionEnabled = "companion_enabled"
        case idleSurface = "idle_surface"
    }
}

struct SloppadCompanionConfigPatch: Encodable {
    let companionEnabled: Bool?
    let idleSurface: String?

    private enum CodingKeys: String, CodingKey {
        case companionEnabled = "companion_enabled"
        case idleSurface = "idle_surface"
    }
}

struct SloppadCompanionStateResponse: Decodable {
    let companionEnabled: Bool
    let idleSurface: String
    let state: String
    let reason: String

    private enum CodingKeys: String, CodingKey {
        case companionEnabled = "companion_enabled"
        case idleSurface = "idle_surface"
        case state
        case reason
    }
}

struct SloppadLivePolicyRequest: Encodable {
    let policy: String
}

enum SloppadCompanionIdleSurface: String, Equatable {
    case robot
    case black

    init(raw: String) {
        self = raw.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() == "black" ? .black : .robot
    }
}

enum SloppadDialogueRuntimeState: String, Equatable {
    case idle
    case listening
    case recording
    case thinking
    case talking
    case error

    init(raw: String) {
        switch raw.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() {
        case "listening":
            self = .listening
        case "recording":
            self = .recording
        case "thinking":
            self = .thinking
        case "talking":
            self = .talking
        case "error":
            self = .error
        default:
            self = .idle
        }
    }
}

struct SloppadDialogueModePresentation: Equatable {
    let isActive: Bool
    let usesBlackScreen: Bool
    let keepScreenAwake: Bool
    let runtimeState: SloppadDialogueRuntimeState
    let primaryLabel: String
    let secondaryLabel: String
    let tapActionLabel: String

    init(
        isActive: Bool,
        isRecording: Bool,
        isAwaitingAssistant: Bool,
        companionEnabled: Bool,
        idleSurface: String,
        runtimeState: String
    ) {
        let normalizedSurface = SloppadCompanionIdleSurface(raw: idleSurface)
        let derivedState: SloppadDialogueRuntimeState
        if !isActive {
            derivedState = .idle
        } else if isRecording {
            derivedState = .recording
        } else if isAwaitingAssistant {
            derivedState = .thinking
        } else {
            let serverState = SloppadDialogueRuntimeState(raw: runtimeState)
            derivedState = serverState == .idle ? .listening : serverState
        }

        self.isActive = isActive
        usesBlackScreen = isActive && normalizedSurface == .black
        keepScreenAwake = usesBlackScreen
        self.runtimeState = derivedState

        switch derivedState {
        case .idle:
            primaryLabel = companionEnabled ? "Ready" : "Disconnected"
            secondaryLabel = "Start dialogue to hand the screen to voice."
            tapActionLabel = "Start dialogue"
        case .listening:
            primaryLabel = "Listening"
            secondaryLabel = "Tap anywhere on the dialogue surface to record."
            tapActionLabel = "Tap to record"
        case .recording:
            primaryLabel = "Recording"
            secondaryLabel = "Tap again to stop and send audio."
            tapActionLabel = "Tap to stop recording"
        case .thinking:
            primaryLabel = "Working"
            secondaryLabel = "Sloppad is processing your last recording."
            tapActionLabel = "Waiting for Sloppad"
        case .talking:
            primaryLabel = "Reply ready"
            secondaryLabel = "Tap to interrupt and start a new recording."
            tapActionLabel = "Tap to record"
        case .error:
            primaryLabel = "Attention needed"
            secondaryLabel = "Check the connection banner for the latest error."
            tapActionLabel = "Tap to retry"
        }
    }
}

struct SloppadAudioCaptureMessage: Encodable {
    let type: String
    let mimeType: String?
    let data: String?

    private enum CodingKeys: String, CodingKey {
        case type
        case mimeType = "mime_type"
        case data
    }
}

struct SloppadInkPoint: Encodable {
    let x: Double
    let y: Double
    let pressure: Double
    let tiltX: Double
    let tiltY: Double
    let roll: Double
    let timestampMS: Double

    private enum CodingKeys: String, CodingKey {
        case x
        case y
        case pressure
        case tiltX = "tilt_x"
        case tiltY = "tilt_y"
        case roll
        case timestampMS = "timestamp_ms"
    }
}

struct SloppadInkStroke: Encodable {
    let pointerType: String
    let width: Double
    let points: [SloppadInkPoint]

    private enum CodingKeys: String, CodingKey {
        case pointerType = "pointer_type"
        case width
        case points
    }
}

struct SloppadInkCommitMessage: Encodable {
    let type: String
    let artifactKind: String
    let requestResponse: Bool
    let outputMode: String
    let totalStrokes: Int
    let strokes: [SloppadInkStroke]

    private enum CodingKeys: String, CodingKey {
        case type
        case artifactKind = "artifact_kind"
        case requestResponse = "request_response"
        case outputMode = "output_mode"
        case totalStrokes = "total_strokes"
        case strokes
    }
}

struct SloppadDiscoveredServer: Identifiable, Hashable {
    let id: String
    let name: String
    let host: String
    let port: Int

    var baseURLString: String {
        "http://\(host):\(port)"
    }
}

func sloppadWSURL(baseURL: URL, path: String) -> URL? {
    guard var components = URLComponents(url: baseURL, resolvingAgainstBaseURL: false) else {
        return nil
    }
    components.scheme = components.scheme == "https" ? "wss" : "ws"
    components.path = "/ws/" + path
    return components.url
}

func sloppadAPIURL(baseURL: URL, path: String) -> URL {
    baseURL.appendingPathComponent("api").appendingPathComponent(path)
}

func sloppadCanvasHTML(from payload: SloppadCanvasEventPayload) -> String {
    if let html = payload.html, html.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty == false {
        return html
    }
    let text = payload.markdownOrText ?? payload.text ?? ""
    let escaped = text
        .replacingOccurrences(of: "&", with: "&amp;")
        .replacingOccurrences(of: "<", with: "&lt;")
        .replacingOccurrences(of: ">", with: "&gt;")
    return "<pre style=\"white-space: pre-wrap; font: -apple-system-body; margin: 24px;\">\(escaped)</pre>"
}

File: platforms/android/app/src/main/kotlin/com/sloppad/android/SloppadModels.kt
package com.sloppad.android

import org.json.JSONArray
import org.json.JSONObject
import java.net.URI
import java.net.URLEncoder
import java.nio.charset.StandardCharsets
import java.util.Base64

data class SloppadWorkspaceListResponse(
    val activeWorkspaceId: String,
    val workspaces: List<SloppadWorkspace>,
)

data class SloppadWorkspace(
    val id: String,
    val name: String,
    val rootPath: String,
    val chatSessionId: String,
    val canvasSessionId: String,
)

data class SloppadRenderedMessage(
    val id: String,
    val role: String,
    val text: String,
    val html: String = "",
)

data class SloppadCanvasArtifact(
    val kind: String,
    val title: String,
    val html: String,
    val text: String,
)

data class SloppadChatEventPayload(
    val type: String,
    val turnId: String = "",
    val role: String = "",
    val message: String = "",
    val markdown: String = "",
    val html: String = "",
    val error: String = "",
    val text: String = "",
    val reason: String = "",
    val state: String = "",
    val workspacePath: String = "",
    val actionType: String = "",
)

data class SloppadCompanionConfig(
    val companionEnabled: Boolean,
    val idleSurface: String,
)

data class SloppadCompanionState(
    val companionEnabled: Boolean,
    val idleSurface: String,
    val state: String,
    val reason: String,
)

enum class SloppadCompanionIdleSurface(val wireValue: String) {
    ROBOT("robot"),
    BLACK("black");

    companion object {
        fun normalize(raw: String): SloppadCompanionIdleSurface {
            return if (raw.trim().lowercase() == BLACK.wireValue) BLACK else ROBOT
        }
    }
}

enum class SloppadDialogueRuntimeState {
    IDLE,
    LISTENING,
    RECORDING,
    THINKING,
    TALKING,
    ERROR;

    companion object {
        fun normalize(raw: String): SloppadDialogueRuntimeState {
            return when (raw.trim().lowercase()) {
                "listening" -> LISTENING
                "recording" -> RECORDING
                "thinking" -> THINKING
                "talking" -> TALKING
                "error" -> ERROR
                else -> IDLE
            }
        }
    }
}

data class SloppadDialogueModePresentation(
    val isActive: Boolean,
    val isRecording: Boolean,
    val isAwaitingAssistant: Boolean,
    val companionEnabled: Boolean,
    val idleSurface: String,
    val runtimeStateValue: String,
) {
    val effectiveIdleSurface = SloppadCompanionIdleSurface.normalize(idleSurface)
    val usesBlackScreen = isActive && effectiveIdleSurface == SloppadCompanionIdleSurface.BLACK
    val keepScreenAwake = usesBlackScreen
    val runtimeState = when {
        !isActive -> SloppadDialogueRuntimeState.IDLE
        isRecording -> SloppadDialogueRuntimeState.RECORDING
        isAwaitingAssistant -> SloppadDialogueRuntimeState.THINKING
        else -> SloppadDialogueRuntimeState.normalize(runtimeStateValue).let {
            if (it == SloppadDialogueRuntimeState.IDLE) SloppadDialogueRuntimeState.LISTENING else it
        }
    }
    val primaryLabel = when (runtimeState) {
        SloppadDialogueRuntimeState.IDLE -> if (companionEnabled) "Ready" else "Disconnected"
        SloppadDialogueRuntimeState.LISTENING -> "Listening"
        SloppadDialogueRuntimeState.RECORDING -> "Recording"
        SloppadDialogueRuntimeState.THINKING -> "Working"
        SloppadDialogueRuntimeState.TALKING -> "Reply ready"
        SloppadDialogueRuntimeState.ERROR -> "Attention needed"
    }
    val secondaryLabel = when (runtimeState) {
        SloppadDialogueRuntimeState.IDLE -> "Start dialogue to hand the screen to voice."
        SloppadDialogueRuntimeState.LISTENING -> "Tap anywhere on the dialogue surface to record."
        SloppadDialogueRuntimeState.RECORDING -> "Android keeps the foreground mic service active while recording."
        SloppadDialogueRuntimeState.THINKING -> "Sloppad is processing your last recording."
        SloppadDialogueRuntimeState.TALKING -> "Tap to interrupt and start a new recording."
        SloppadDialogueRuntimeState.ERROR -> "Check the connection banner for the latest error."
    }
    val tapActionLabel = when (runtimeState) {
        SloppadDialogueRuntimeState.IDLE -> "Start dialogue"
        SloppadDialogueRuntimeState.LISTENING -> "Tap to record"
        SloppadDialogueRuntimeState.RECORDING -> "Tap to stop recording"
        SloppadDialogueRuntimeState.THINKING -> "Waiting for Sloppad"
        SloppadDialogueRuntimeState.TALKING -> "Tap to record"
        SloppadDialogueRuntimeState.ERROR -> "Tap to retry"
    }
}

data class SloppadInkPoint(
    val x: Float,
    val y: Float,
    val pressure: Float,
    val tiltX: Float,
    val tiltY: Float,
    val roll: Float,
    val timestampMs: Long,
)

data class SloppadInkStroke(
    val pointerType: String,
    val width: Float,
    val points: List<SloppadInkPoint>,
)

data class SloppadDiscoveredServer(
    val id: String,
    val name: String,
    val host: String,
    val port: Int,
) {
    val baseUrlString: String
        get() = "http://$host:$port"
}

fun sloppadWsUrl(baseUrl: String, path: String): String {
    val base = URI(baseUrl.trim())
    val scheme = if (base.scheme.equals("https", ignoreCase = true)) "wss" else "ws"
    val authority = base.rawAuthority ?: error("base URL is missing an authority: $baseUrl")
    val encodedPath = path
        .split("/")
        .joinToString("/") { segment -> URLEncoder.encode(segment, StandardCharsets.UTF_8).replace("+", "%20") }
    return "$scheme://$authority/ws/$encodedPath"
}

fun sloppadApiUrl(baseUrl: String, path: String): String {
    return "${baseUrl.trim().trimEnd('/')}/api/$path"
}

fun parseWorkspaceListResponse(body: String): SloppadWorkspaceListResponse {
    val json = JSONObject(body)
    val workspaces = buildList {
        val items = json.optJSONArray("workspaces") ?: JSONArray()
        for (index in 0 until items.length()) {
            val item = items.optJSONObject(index) ?: continue
            add(
                SloppadWorkspace(
                    id = item.optString("id"),
                    name = item.optString("name"),
                    rootPath = item.optString("root_path"),
                    chatSessionId = item.optString("chat_session_id"),
                    canvasSessionId = item.optString("canvas_session_id"),
                )
            )
        }
    }
    return SloppadWorkspaceListResponse(
        activeWorkspaceId = json.optString("active_workspace_id"),
        workspaces = workspaces,
    )
}

fun parseChatHistory(body: String): List<SloppadRenderedMessage> {
    val json = JSONObject(body)
    val messages = json.optJSONArray("messages") ?: JSONArray()
    return buildList {
        for (index in 0 until messages.length()) {
            val item = messages.optJSONObject(index) ?: continue
            val markdown = item.optString("content_markdown")
            val plain = item.optString("content_plain")
            add(
                SloppadRenderedMessage(
                    id = "persisted-${item.optLong("id")}",
                    role = item.optString("role"),
                    text = markdown.takeIf { it.isNotBlank() } ?: plain,
                )
            )
        }
    }
}

fun parseCanvasSnapshot(body: String): SloppadCanvasArtifact? {
    val event = JSONObject(body).optJSONObject("event") ?: return null
    return parseCanvasArtifact(event)
}

fun parseCanvasArtifact(payload: JSONObject): SloppadCanvasArtifact {
    val text = payload.optString("text").ifBlank { payload.optString("markdown_or_text") }
    return SloppadCanvasArtifact(
        kind = payload.optString("kind"),
        title = payload.optString("title"),
        html = payload.optString("html").ifBlank { wrapCanvasText(text) },
        text = text,
    )
}

fun parseChatEvent(raw: String): SloppadChatEventPayload {
    val json = JSONObject(raw)
    val action = json.optJSONObject("action")
    return SloppadChatEventPayload(
        type = json.optString("type"),
        turnId = json.optString("turn_id"),
        role = json.optString("role"),
        message = json.optString("message"),
        markdown = json.optString("markdown"),
        html = json.optString("html"),
        error = json.optString("error"),
        text = json.optString("text"),
        reason = json.optString("reason"),
        state = json.optString("state"),
        workspacePath = json.optString("workspace_path"),
        actionType = action?.optString("type").orEmpty(),
    )
}

fun parseCompanionConfig(body: String): SloppadCompanionConfig {
    val json = JSONObject(body)
    return SloppadCompanionConfig(
        companionEnabled = json.optBoolean("companion_enabled"),
        idleSurface = json.optString("idle_surface", SloppadCompanionIdleSurface.ROBOT.wireValue),
    )
}

fun parseCompanionState(body: String): SloppadCompanionState {
    val json = JSONObject(body)
    return SloppadCompanionState(
        companionEnabled = json.optBoolean("companion_enabled"),
        idleSurface = json.optString("idle_surface", SloppadCompanionIdleSurface.ROBOT.wireValue),
        state = json.optString("state"),
        reason = json.optString("reason"),
    )
}

fun loginRequest(password: String): String {
    return JSONObject().put("password", password).toString()
}

fun composerRequest(text: String): String {
    return JSONObject()
        .put("text", text)
        .put("output_mode", "voice")
        .toString()
}

fun companionConfigPatch(companionEnabled: Boolean? = null, idleSurface: String? = null): String {
    val json = JSONObject()
    if (companionEnabled != null) {
        json.put("companion_enabled", companionEnabled)
    }
    if (!idleSurface.isNullOrBlank()) {
        json.put("idle_surface", idleSurface)
    }
    return json.toString()
}

fun livePolicyRequest(policy: String): String {
    return JSONObject().put("policy", policy).toString()
}

fun audioPcmMessage(data: ByteArray): String {
    return JSONObject()
        .put("type", "audio_pcm")
        .put("mime_type", "audio/L16;rate=16000;channels=1")
        .put("data", Base64.getEncoder().withoutPadding().encodeToString(data))
        .toString()
}

fun audioStopMessage(): String {
    return JSONObject().put("type", "audio_stop").toString()
}

fun inkCommitMessage(strokes: List<SloppadInkStroke>, requestResponse: Boolean): String {
    val items = JSONArray()
    for (stroke in strokes) {
        val points = JSONArray()
        for (point in stroke.points) {
            points.put(
                JSONObject()
                    .put("x", point.x)
                    .put("y", point.y)
                    .put("pressure", point.pressure)
                    .put("tilt_x", point.tiltX)
                    .put("tilt_y", point.tiltY)
                    .put("roll", point.roll)
                    .put("timestamp_ms", point.timestampMs)
            )
        }
        items.put(
            JSONObject()
                .put("pointer_type", stroke.pointerType)
                .put("width", stroke.width)
                .put("points", points)
        )
    }
    return JSONObject()
        .put("type", "ink_stroke")
        .put("artifact_kind", "text")
        .put("request_response", requestResponse)
        .put("output_mode", "voice")
        .put("total_strokes", strokes.size)
        .put("strokes", items)
        .toString()
}

private fun wrapCanvasText(text: String): String {
    val escaped = text
        .replace("&", "&amp;")
        .replace("<", "&lt;")
        .replace(">", "&gt;")
    return "<pre style=\"white-space: pre-wrap; margin: 24px; font: sans-serif;\">$escaped</pre>"
}
